avr-hal-generic/src/spi.rs

```diff
@@ -299,3 +303,63 @@ macro_rules! impl_spi {
         }
     };
 }
+
+
+/// Implement traits for a SPI interface
+#[macro_export]
+macro_rules! impl_spi {
```

Comment:
I think it should be possible to do this without introducing a second macro if `impl_spi!` just invokes itself:

```rust
#[macro_export]
macro_rules! impl_spi {
    (
        $(#[$spi_attr:meta])*
        pub struct $Spi:ident {
            peripheral: $SPI:ty,
            pins: {
                sclk: $sclkmod:ident::$SCLK:ident,
                mosi: $mosimod:ident::$MOSI:ident,
                miso: $misomod:ident::$MISO:ident,
                cs: $csmod:ident::$CS:ident,
            }
        }
    ) => {
        $crate::impl_spi! {
            pub struct $Spi {
                peripheral: $SPI,
                pins: {
                    sclk: $sclkmod::$SCLK,
                    mosi: $mosimod::$MOSI,
                    miso: $misomod::$MISO,
                    cs: $csmod::$CS,
                }
            }
            pub struct ChipSelectPin;
        }
    };
    (
        $(#[$spi_attr:meta])*
        pub struct $Spi:ident {
            peripheral: $SPI:ty,
            pins: {
                sclk: $sclkmod:ident::$SCLK:ident,
                mosi: $mosimod:ident::$MOSI:ident,
                miso: $misomod:ident::$MISO:ident,
                cs: $csmod:ident::$CS:ident,
            }
        }
        pub struct $ChipSelectPin:ident;
    )
    => {
        // actual implementation ...
    };
}
```

Reply:
wow, this is clever. I'll try it.